metaflow/client/cache/cache_store.py

```diff
@@ -0,0 +1,260 @@
+
+        # It is possible that the datastore contains more than gc_watermark
+        # bytes. To ensure that we start below the gc_watermark, we run the GC:
+        self._gc_objects()
+        # the above method may have marked objects for deletion. We assume that
+        # no clients are accessing the objects since the server hasn't started
+        # yet, so we can safely delete the marked objects without a quarantine:
+        self._gc_objects(quarantine=-1)
+
+        self.echo("Cache initialized with %d permanents objects, "\
+                  "%d disposable objects, totaling %d bytes."\
+                  % (len(self.objects_queue),
+                     len(self.disposables_queue),
+                     self.total_size))
+
+    def _gc_objects(self, quarantine=GC_MARKER_QUARANTINE):
+
+        def mark_for_deletion(path, size):
+            self.safe_fileop(os.utime, path, (TIMESTAMP_FOR_DELETABLE,
+                                              TIMESTAMP_FOR_DELETABLE))
+            self.gc_queue[path] = (time.time(), size)
+
+        # 1) delete marked objects that are past their quarantine period
+        limit = time.time() - quarantine
+        deleted = []
+        for path in list(self.gc_queue):
+            tstamp, size = self.gc_queue[path]
+            if tstamp < limit:
+                if self.safe_fileop(os.unlink, path):
+                    del self.gc_queue[path]
+                    self.total_size -= size
+            else:
+                break
+
+        # if there are still objects marked for deletion, we can just wait
+        # for them to age past quarantine. Without this check, we could GC
+        # objects too eagerly during the quarantine period.
+        if not self.gc_queue:
+
+            # 2) if we are still using too much space, mark disposable
+            # objects for deletion
+            unmarked_size = self.total_size
+            while self.disposables_queue and unmarked_size > self.gc_watermark:
+                path, size = self.disposables_queue.popitem(last=False)
+                mark_for_deletion(path, size)
+                unmarked_size -= size
+
+            # 3) after we have exhausted all disposables, we need to start
+            # marking non-disposable objects for deletion, starting from the
+            # oldest.
+            while self.objects_queue and unmarked_size > self.gc_watermark:
+                path, size = self.objects_queue.popitem(last=False)
+                mark_for_deletion(path, size)
+                unmarked_size -= size
+
+    def ensure_path(self, path):
+        dirr = os.path.dirname(path)
+        if not os.path.isdir(dirr):
+            try:
+                makedirs(dirr)
+            except Exception as ex:
+                self.warn(ex, "Could not create dir: %s" % path)
+
+    def open_tempdir(self, token, action_name, stream_key):
+        self._gc_objects()
+        if self.total_size > self.max_size:
```

Comment:
So to be clear, the cache can go higher than that size because the objects are marked for deletion but not deleted, correct? Further, an action can just fill up this space easily before commit and it wouldn't count.